listener: add name to filter chains

kailun-qin opened the pull request:
This patch introduces the generation of a unique logical name for unnamed
filter chains. This can help some debugging of errors like `error adding
listener '0.0.0.0:8443': filter chain '' has the same matching rules defined
as ''`.

Fixes https://github.com/istio/istio/issues/27480

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

istio-testing commented:
Hi @kailun-qin. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

ramaraochavali commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -2041,7 +2045,11 @@ func buildCompleteFilterChain(mutable *istionetworking.MutableObjects, opts buil
 		mutable.Listener.FilterChains[i].Metadata = opt.metadata
-		mutable.Listener.FilterChains[i].Name = opt.filterChainName
+		if len(opt.filterChainName) == 0 {
+			mutable.Listener.FilterChains[i].Name = FilterChainNamePrefix + uuid.New().String()
```
Would it be useful if we add listener name to the filter chain name? and if we add it just append incrementing counter for each fc instead of generating uuid?

kailun-qin replied:
> Would it be useful if we add listener name to the filter chain name? and if we add it just append incrementing counter for each fc instead of generating uuid?

Thanks a for the comments!
My two cents:
1) I think it would be somehow useful to add listener name to the filter chain name.
2) As for the id generation, IMHO, I would prefer uuid to increment counter if the key concern here is not performance. Using increment counter might bring more concerns to e.g. stateful chain name, security (info/status disclosure) and  restarted pilot etc.

kailun-qin replied:
Anyway, let's see if there's further input from the community and I'm open to it.

howardjohn replied:
I think ideally we are adding a filter chain name to the real generate filterchains. So its useful in real world, not just our tests. Note it must be deterministic then

hzxuzhonghu replied:
The filter chain names can not be same, so i think this is why not set at beginning.

kailun-qin commented:
Thank you all for your comments! Updated a new version.

Using <listener_name>-fc-<uuid> as the generated filter chain name. This should be 
* deterministic (that can be applied to real generate filter chains, since Envoy would do a UUID-like way for empty names according to its [proto/api](https://github.com/envoyproxy/envoy/blob/18db4c90e3295fb2c39bfc7b2ce641cfd6c3fbed/api/envoy/config/listener/v3/listener_components.proto#L253)) 
* and also added w/ listener name (as prefix) to help w/ easy identification.

howardjohn commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -2041,7 +2045,11 @@ func buildCompleteFilterChain(mutable *istionetworking.MutableObjects, opts buil
 		mutable.Listener.FilterChains[i].Metadata = opt.metadata
-		mutable.Listener.FilterChains[i].Name = opt.filterChainName
+		if len(opt.filterChainName) == 0 {
+			mutable.Listener.FilterChains[i].Name = mutable.Listener.Name + FilterChainNameInfix + uuid.New().String()
```
sorry, I don't understand how a UUID can every be deterministic? The next time we re-generate listeners the name will be different and envoy will do a full drain on the listener, right?

kailun-qin replied:
Right, sorry I misunderstood. I may instead use a hash on some fields of the filterchain as the uid. What do you think?

irisdingbj commented:
/ok-to-test

istio-testing commented:
@kailun-qin: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1354982444796743680) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1354982444796743680) | `/test release-notes_istio`
lint_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1354982443110633472) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1354982444796743680) | `/test release-notes_istio`
lint_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1354982443110633472) | `/test lint_istio`
unit-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/unit-tests_istio/1354982438916329472) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1354982444796743680) | `/test release-notes_istio`
lint_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1354982443110633472) | `/test lint_istio`
unit-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/unit-tests_istio/1354982438916329472) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/integ-pilot-k8s-tests_istio/1354982438962466816) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1354982443110633472) | `/test lint_istio`
unit-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/unit-tests_istio/1354982438916329472) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/integ-pilot-k8s-tests_istio/1354982438962466816) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 724ed8f799c605f9c158bab3ff7335fe63dba1bd | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1355001081465147392) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/unit-tests_istio/1354982438916329472) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | adf5718bdce7c60267c71142a1f51cd0811c21d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/integ-pilot-k8s-tests_istio/1354982438962466816) | `/test integ-pilot-k8s-tests_istio`
release-notes_istio | 724ed8f799c605f9c158bab3ff7335fe63dba1bd | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1355001081465147392) | `/test release-notes_istio`
lint_istio | 724ed8f799c605f9c158bab3ff7335fe63dba1bd | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1355001079783231488) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 724ed8f799c605f9c158bab3ff7335fe63dba1bd | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1355001079783231488) | `/test lint_istio`
release-notes_istio | be1f058500eb728910ccbe478c2ff571710e865c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1355011666382688256) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | be1f058500eb728910ccbe478c2ff571710e865c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1355011666382688256) | `/test release-notes_istio`
lint_istio | be1f058500eb728910ccbe478c2ff571710e865c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1355011664700772352) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kailun-qin: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | be1f058500eb728910ccbe478c2ff571710e865c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1355011664700772352) | `/test lint_istio`
release-notes_istio | fcc4c57a2aa8521094c60e70a5c3b78a347fa63a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/release-notes_istio/1355013116630405120) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kailun-qin commented:
Updated a new version using <listener_name>-fc-<hash of the unnamed filterchain> as the identifier.

kailun-qin commented:
> Updated a new version using <listener_name>-fc-<unnamed_filterchain_hash> as the identifier.

Would just like to follow up, any comment on the updated approach? @howardjohn Thanks a lot!

howardjohn commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -2024,6 +2028,17 @@ func (configgen *ConfigGeneratorImpl) appendListenerFallthroughRouteForCompleteL
 
+// hash returns a unique hash based on listener.FilterChain, default to 32-bit FNV-1a hash for now
+func hash(fc *listener.FilterChain) (res string, err error) {
+	h := fnv.New32a()
```
I have some concerns about the cost and potential non-determinism of the hash. The risk and cost here may outweigh the benefits of getting a name.

ramaraochavali replied:
I share same concern about cost. Can we run some benchmark to see what is the overhead if any?

kailun-qin replied:
Many thanks for your comments!

IMHO, a benchmark might help but I guess it highly depends on the size of `listener.FilterChain`. And I would agree w/ @howardjohn to some extent, though I think hash can be deterministic if no map-like inputs involved and the cost is OK since it's a general practice.

Besides using a complete standard hash on `listener.FilterChain` for deterministic naming, we have other options like:
1) using a `logical` hash which concatenates some fields within filterchian. However, it's somewhat difficult to decide which fields to pick up as some might be empty; the logical name generated may have implicit semantics which can be confusing; and if we'd like to eliminate that confusion, we'd take another hash anyway.
2) hash on parts of the fields within filterchain. This decreases the cost (though still introduced) but increases the possibility that multi filterchains can get the same name (though seems acceptable for Envoy but loses the intent to be distinct).

Otherwise, if we do have concerns about the cost, can we make this auto-generation configurable similarly to what has been applied in kubectl configmap/secret https://github.com/kubernetes/kubernetes/commit/503a6a8eecd1cbaeceaf54c78ebac14d4d89a4b0?

kailun-qin commented:
Added benchmark tests with some perf optimizations. Please kindly see the results below.
@ramaraochavali @howardjohn Any comments? Thanks!
```
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/networking/core/v1alpha3
BenchmarkFilterChainHash
BenchmarkFilterChainHash/virtualOutbound-blackhole
BenchmarkFilterChainHash/virtualOutbound-blackhole-16             947964              1165 ns/op
BenchmarkFilterChainHash/virtualOutbound-catchall-tcp
BenchmarkFilterChainHash/virtualOutbound-catchall-tcp-16          893870              1353 ns/op
BenchmarkFilterChainHash/virtualInbound-blackhole
BenchmarkFilterChainHash/virtualInbound-blackhole-16             1000000              1155 ns/op
BenchmarkFilterChainHash/virtualInbound
BenchmarkFilterChainHash/virtualInbound-16                        499588              2502 ns/op
BenchmarkFilterChainHash/virtualInbound#01
BenchmarkFilterChainHash/virtualInbound#01-16                     600506              1919 ns/op
BenchmarkFilterChainHash/virtualInbound-catchall-http
BenchmarkFilterChainHash/virtualInbound-catchall-http-16          337414              3640 ns/op
BenchmarkFilterChainHash/virtualInbound-catchall-http#01
BenchmarkFilterChainHash/virtualInbound-catchall-http#01-16       416407              3120 ns/op
```

istio-testing commented:
@kailun-qin: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 52ec8694cb836e190dd3c97f6e8b3b92cf29c41b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30433/lint_istio/1363758711419441152) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -2024,6 +2029,22 @@ func (configgen *ConfigGeneratorImpl) appendListenerFallthroughRouteForCompleteL
+
+	h := crc32.New(crc32.MakeTable(crc32.Castagnoli))
+
+	if fcJSON, _ = json.Marshal(fc); err != nil {
```
json marshaling the whole filter chain is not acceptable for performance. We used to do this, and it added a 10x overhead. Since we are only getting basic debugging improvement from this, I don't think its worth it.

kailun-qin replied:
Thanks for the feedbacks!
Though comparing w/ other struct -> string solution, json marshaling has relatively better performance. Anyway, I would agree w/ the performance concerns and would like to close this one if no better solution.
